### PR TITLE
Networking Improvements

### PR DESCRIPTION
**Pull request by kontsevoy**
## Problem 1: Idle Connections

Up to this day Teleport did not use read/write timeouts for SSH connections. This can lead to frozen connections. A nicer connection handling would set an "idle timeout" on every connection which means it will be dropped after a certain idle interval passes.

## Problem 2: Real Client IP

For audit purposes it is often required to log the IP of every connected SSH client. However, if an SSH bastion is used, it's the address of the bastion what gets logged as "client". We need to find a way to log the real actual client IP into the audit log.

## Changes

There are 3 commits here:

1. Refactoring: added comments, renamed a few things, eliminated dead code. Preparing Teleport for a new developer. ;) Commit 2035ace86
2. Idle connection handling. Commit ed8604f75
3. True client IP forwarding. Commit 4f009e225 (more on this below)

## Client IP Forwarding Algorithm

There is no standard "SSH way" to forward the true client IP to a server. The initial conditions are:
* The client does not know its own IP, it's almost always behind NAT.
* Only proxy/bastion knows the external client IP.
* But the proxy does not have the credentials to communicate this information to the SSH server (proxy is a classic unauthenticated man-in-the-middle here).
* We also need to make sure we don't break compatibility with OpenSSH clients and OpenSSH servers.

The implemented algo works by proxy injecting a JSON payload in the middle of "SSH version Exchange" (see corresponding [RFC](https://tools.ietf.org/html/rfc4253#page-4))

The version exchange with a Teleport server is shown below as seen on the wire. It begins by server sending its version string first:
```
server-to-proxy > SSH-2.0-Teleport
proxy-to-client > SSH-2.0-Teleport
client-to-proxy > SSH-2.0-Client
proxy-to-server > SSH-2.0-Client + JSON payload
```

and here's what happens with OpenSSH server (or with old Teleport versions):
```
server-to-proxy > SSH-2.0-Go
proxy-to-client > SSH-2.0-Go
client-to-proxy > SSH-2.0-Client
proxy-to-server > SSH-2.0-Client
```

in other words, when a proxy sees that a server sends "SSH-2.0-Teleport" version string, it _knows_ that the server is capable of receiving a custom JSON payload right after the SSH version exchange.

### CODE REVIEW

File: lib/srv/proxy.go [hunk 1]
```diff
@@ -276,3 +293,39 @@ func (t *proxySubsys) wait() error {
...
+	}
+	// is that a Teleport server?
+	resp := bytes[:n]
+	if strings.HasPrefix(string(resp), sshutils.SSHVersionPrefix) {
+		// if we're connecting to a Teleport SSH server, send our own "handshake payload"
+		// message, along with a client's IP:
+		hp := &sshutils.HandshakePayload{
+			ClientAddr: clientAddr.String(),
+		}
+		hpJson, err := json.Marshal(hp)
```
Comment (klizhentas) [inline lib/srv/proxy.go]: `hpJson` -> `hpJSON`

File: lib/sshutils/server.go [hunk 2]
```diff
@@ -297,3 +342,85 @@ func KeysEqual(ak, bk ssh.PublicKey) bool {
...
+
+// RemoteAddr returns the behind-the-proxy client address
+func (c *connectionWrapper) RemoteAddr() net.Addr {
+	return c.clientAddr
+}
+
+// Read implements io.Read() part of net.Connection which allows the
+// connection wrapper to peek at the beginning of SSH handshake
+func (c *connectionWrapper) Read(b []byte) (int, error) {
+	if c.upstreamReader == nil {
```
Comment (klizhentas) [inline lib/sshutils/server.go]: to reduce nesting, let's invert this:

```go
if c.upstreamReader ! = nil {
    return ...
}
... rest of code here
}

**Comment by kontsevoy** [comment]
retest this please
